File: src/main/java/de/fraunhofer/aisec/cpg/graph/types/WrapState.java
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types;

/** Stores State for rewrap when typeinformation has been unwrapped */
public class WrapState {

  int depth;
  boolean reference;
  PointerType.PointerOrigin pointerOrigin;
  ReferenceType referenceType;

  public WrapState() {
    this.depth = 0;
    this.reference = false;
    this.pointerOrigin = PointerType.PointerOrigin.ARRAY;
    this.referenceType = null;
  }

  public int getDepth() {
    return depth;
  }

  public void setDepth(int depth) {
    this.depth = depth;
  }

  public boolean isReference() {
    return reference;
  }

  public void setReference(boolean reference) {
    this.reference = reference;
  }

  public PointerType.PointerOrigin getPointerOrigin() {
    return pointerOrigin;
  }

  public void setPointerOrigin(PointerType.PointerOrigin pointerOrigin) {
    this.pointerOrigin = pointerOrigin;
  }

  public ReferenceType getReferenceType() {
    return referenceType;
  }

  public void setReferenceType(ReferenceType referenceType) {
    this.referenceType = referenceType;
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/types/WrapState.kt
/*
 * Copyright (c) 2021, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types

import java.util.*

/**
 * A [WrapState] holds information how a "wrapped" type (for example a [PointerType]) is built from
 * its element types(s). This can potentially be a chain of different pointer/array operations.
 */
class WrapState {

    /**
     * @param depth The total depth of "wrapping". This is usually equal to
     *   [SecondOrderType.referenceDepth]
     */
    constructor(depth: Int = 0) {
        wraps = arrayOfNulls(depth)
    }

    /** An array of [Wrap] values, applied in the order the types are wrapped in. */
    var wraps: Array<Wrap?>

    enum class Wrap {
        ARRAY,
        POINTER,
        REFERENCE,
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is WrapState) return false

        return wraps.contentEquals(other.wraps)
    }

    override fun hashCode(): Int {
        return wraps.contentHashCode()
    }
}
